time/time.c

```diff
@@ -155,59 +155,102 @@ double difftime(time_t t1, time_t t2)
 }
 
 
-struct tm *gmtime_r(const time_t *timep, struct tm *res)
+static int epochDaysToYears(int days, int *yearDay)
 {
-	time_t sec = *timep, minute = 0, hou = 0, t;
-	int day = 0, leap, leapcnt;
+	/* Move day 0 from 1970-01-01 to 2000-02-29. 2000-02-29 closest date when all periods restart.
+	 * Algorithm also works correctly when days < 0.
+	 */
+	int year = 2000, leapDay, fullPeriods;
+	days -= 11016;
+
+	static const struct {
+		int nDays;         /* Number of days this period has */
+		int nYears;        /* Number of years this period has */
+		int isZeroDayLeap; /* Is day 0 of this period a leap day (unless a larger period was tried before) */
+	} periods[4] = {
+		{ 146097, 400, 1 },
+		{ 36524, 100, 0 },
+		{ 1461, 4, 1 },
+		{ 365, 1, 0 },
+	};
+
+	for (int i = 0; i < 4; i++) {
+		fullPeriods = days / periods[i].nDays;
+		days = days % periods[i].nDays;
+		if (days < 0) {
+			days += periods[i].nDays;
+			fullPeriods -= 1;
+		}
 
-	if ((t = sec % 60) != sec) {
-		minute = sec / 60;
-		sec = t;
+		year += fullPeriods * periods[i].nYears;
+		leapDay = periods[i].isZeroDayLeap;
+		if (days == 0) {
+			break;
+		}
 	}
 
-	if ((t = minute % 60) != minute) {
-		hou = minute / 60;
-		minute = t;
+	/* Day 0 is now Feb 29th if leapDay == 1, otherwise it's Feb 28th (leap years have two 0-days!)
+	 * Day 1 is always March 1st, but which year day depends on if the year is a leap year.
+	 * Day 307 and higher are next year, but before Feb 28th (before possible leap day, so leap year or not doesn't matter)
+	 */
+	if (days >= 307) {
+		year += 1;
+		*yearDay = days - 307;
 	}
-
-	if ((t = hou % 24) != hou) {
-		day = hou / 24;
-		hou = t;
+	else if (days == 0) {
+		*yearDay = 58 + leapDay;
 	}
+	else {
+		*yearDay = days + (isleap(year) ? 59 : 58);
+	}
+
+	return year;
+}
 
-	res->tm_sec = sec;
-	res->tm_min = minute;
-	res->tm_hour = hou;
 
-	res->tm_wday = (day + 4) % 7;
+struct tm *gmtime_r(const time_t *timep, struct tm *res)
+{
+	time_t seconds = *timep, minutes = 0, hours = 0, t;
+	int days = 0, year, isYearLeap;
```

Comment:
Nitpick - `int` days and year still can be overflowed by `time_t`. I see no point in handling times hundred of thousands of years into the future and past, so perhaps some check to reject too big values?

Reply:
It's a good idea, other parts of the system can't handle timestamps that large anyway (for example timer counts the number of microseconds in a 64-bit variable, so `time_t` generated by it will be 45 bits max). Using signed 32-bit int allows us to handle time_t that are about 46 bits long.